### ISSUE TITLE
partial lazy evaluation (conjunctions, disjunctions)

### ISSUE THREAD

**Issue by witwicki** [aseba-community/aseba#301]
This is not a critical issue, but rather something aseba is lacking that I seem to take for granted when coding in other high-level languages...

In evaluating a conjuction of logical expressions, aseba evaluates all expressions even if the first one is false.

Example:

```
var a[3]
var i = 0
while (i<3 and a[i]==0) do
   i = i + 1
end
```

I would expect this loop to terminate with i set to 3, but instead it dies because "a[i]==0" still gets evaluated even after i<3 becomes false.

Of course, there are simple workarounds.   But this was still a nuisance when I encountered it because I did not expect aseba to behave as such.

**Comment by ypiguet-epfl** [comment]
I disagree with your analysis. With your example, my disassembler gives this for the compiled code (tested in 1.6):
```
    7  3000       load a
    8  1000       push.s 0
    9  800a       eq
   10  3001       load b
   11  1000       push.s 0
   12  800a       eq
   13  a011 xxxx  jump if not and @end
...
@end:

```
It would be the same size, and one instruction less, with lazy evaluation:
```
    7  3000       load a
    8  1000       push.s 0
    9  a00a xxxx  jump if not eq @end
   11  3001       load b
   12  1000       push.s 0
   13  a00a xxxx  jump if not eq @end
...
@end:

```
For the execution speed, lazy evaluation is beneficial most of the time, even in cases it isn't required. In a VM like Aseba's, it's certainly irrelevant.
